### PR TITLE
Allocate type

### PR DESCRIPTION
**Pull request by arporter**
This PR is for #76 - handle allocates using names of derived types as type specifiers.

### CODE REVIEW

**Comment by arporter** [comment]
I've requested a review from either @rupertford or @TeranIvy. Feel free to discuss amongst yourselves which of you is best to do it :-) The reviewing guidelines are here: https://github.com/stfc/fparser/wiki/CodeReview

**Comment by TeranIvy** [comment]
Hppy to review it if @rupertford agrees :-)

File: src/fparser/statements.py [hunk 1]
```diff
@@ -641,13 +642,21 @@ class Allocate(Statement):
...
+        :raises ParseError: if an invalid type-specification is used
+        '''
         line = self.item.get_line()[8:].lstrip()[1:-1].strip()
         item2 = self.item.copy(line, True)
         line2 = item2.get_line()
         i = line2.find('::')
         if i != -1:
             spec = item2.apply_map(line2[:i].rstrip())
             from .block_statements import type_spec
+            from .utils import is_name
```
Comment (TeranIvy) [inline src/fparser/statements.py]: This returns pylint error as from .utils import is_name is already defined at the top of file. Please remove.

File: src/fparser/statements.py [hunk 2]
```diff
@@ -656,8 +665,13 @@ def process_item(self):
...
             if stmt is not None and stmt.isvalid:
                 spec = stmt
+            elif is_name(spec):
+                # Type spec is the name of a derived type
+                pass
             else:
-                self.warning('TODO: unparsed type-spec' + repr(spec))
+                raise ParseError(
+                    "Unrecognised type-specification in ALLOCATE statement: "
+                    "{0}".format(self.item.line))
```
Comment (TeranIvy) [inline src/fparser/statements.py]: Changed warning to ParseError, nice!

File: src/fparser/statements.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/fparser/statements.py]: `pylint` complains about two other things (below). However, this seems to be standard behaviour for other classes as well so no changes are required here.
```
W:679, 8: Attribute 'items' defined outside __init__ (attribute-defined-outside-init)
W:678, 8: Attribute 'spec' defined outside __init__ (attribute-defined-outside-init)
```

File: src/fparser/statements.py [hunk 3]
```diff
@@ -641,13 +642,21 @@ class Allocate(Statement):
     match = re.compile(r'allocate\s*\(.*\)\Z', re.I).match
 
     def process_item(self):
+        '''
+        Process the allocate statement and store the various entities being
```
Comment (TeranIvy) [inline src/fparser/statements.py]: For the sake of consistency with other docstrings for this method, please replace "allocate" with "ALLOCATE".

File: src/fparser/statements.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/statements.py]: Fixed the 1st one and leaving the second.

File: src/fparser/statements.py (hunk 3 above)
Comment (arporter) [reply src/fparser/statements.py]: Done.

**Comment by arporter** [comment]
Ready for second review. I've also just given @TeranIvy admin privileges!

File: src/fparser/tests/test_parser.py [hunk 4]
```diff
@@ -221,12 +222,23 @@ def test_contains():
 
 
 def test_allocate():
+    ''' Tests for various form of allocate '''
```
Comment (TeranIvy) [inline src/fparser/tests/test_parser.py]: Sorry, forgot to ask to capitalise "ALLOCATE" here as well!

File: src/fparser/statements.py [hunk 5]
```diff
@@ -666,11 +679,22 @@ def process_item(self):
         return
 
     def tofortran(self, isfix=None):
-        t = ''
+        '''
+        Create the Fortran for this ALLOCATE statement
```
Comment (TeranIvy) [inline src/fparser/statements.py]: Perhaps "Create the Fortran code..." or "Create the Fortran representation..."?

File: src/fparser/tests/test_parser.py (hunk 4 above)
Comment (arporter) [reply src/fparser/tests/test_parser.py]: Done.

File: src/fparser/statements.py (hunk 5 above)
Comment (arporter) [reply src/fparser/statements.py]: Done.

**Comment by arporter** [comment]
Hi @TeranIvy, thanks for being so thorough! Ready for another review...

**Comment by TeranIvy** [comment]
All changes are done. Proceeding to the merge.

**Comment by TeranIvy** [comment]
Merged to master and deleted the branch.

**Comment by arporter** [comment]
Yes, I wouldn't worry - you only need to check that I've done what you said (and not broken anything in the process).

**Pull request merged**